File: Recount.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2009 Sidoine De Wispelaere.
    Copyright (C) 2012, 2013, 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

local OVALE, Ovale = ...
local OvaleRecount = Ovale:NewModule("OvaleRecount")
Ovale.OvaleRecount = OvaleRecount

--<private-static-properties>
local L = nil
local OvaleScore = nil
local Recount = LibStub("AceAddon-3.0"):GetAddon("Recount", true)
local setmetatable = setmetatable
-- GLOBALS: GameTooltip
-- GLOBALS: LibStub
--</private-static-properties>

--<private-static-methods>
local function DataModes(self, data, num)
	if not data then
		return 0, 0
	end
	local fight = data.Fights[Recount.db.profile.CurDataSet]
	local score
	if fight and fight.Ovale and fight.OvaleMax then
		score = fight.Ovale * 1000 / fight.OvaleMax
	else
		score = 0
	end
	if num == 1 then
		return score
	end
	return score, nil
end

local function TooltipFuncs(self, name, data)
	local SortedData, total
	GameTooltip:ClearLines()
	GameTooltip:AddLine(name)
	-- Recount:AddSortedTooltipData(L["Top 3"].." Ovale",data and data.Fights[Recount.db.profile.CurDataSet] and data.Fights[Recount.db.profile.CurDataSet].Ovale,3)
	-- GameTooltip:AddLine("<"..L["Click for more Details"]..">",0,0.9,0)
end
--</private-static-methods>

--<public-static-methods>
function OvaleRecount:OnInitialize()
	OvaleScore = Ovale.OvaleScore
	if Recount then
		local AceLocale = LibStub("AceLocale-3.0", true)
		local L = AceLocale and AceLocale:GetLocale("Recount", true)
		if not L then
			L = setmetatable({}, { __index = function(t, k) t[k] = k; return k; end })
		end
		Recount:AddModeTooltip(OVALE, DataModes, TooltipFuncs, nil, nil, nil, nil)
	end
end

function OvaleRecount:OnEnable()
	if Recount then
		OvaleScore:RegisterDamageMeter("OvaleRecount", self, "ReceiveScore")
	end
end

function OvaleRecount:OnDisable()
	OvaleScore:UnregisterDamageMeter("OvaleRecount")
end

function OvaleRecount:ReceiveScore(name, guid, scored, scoreMax)
	if Recount then
		local source = Recount.db2.combatants[name]
		if source then
			Recount:AddAmount(source, OVALE, scored)
			Recount:AddAmount(source, OVALE .. "Max", scoreMax)
		end
	end
end
--</public-static-methods>

File: src/Recount.ts
import AceLocale from "@wowts/ace_locale-3.0";
import Recount from "@wowts/recount";
import { setmetatable, LuaObj } from "@wowts/lua";
import { GameTooltip } from "@wowts/wow-mock";
import { OvaleClass } from "./Ovale";
import { OvaleScoreClass } from "./Score";

const DataModes = function(self: never, data: any, num: number) {
    if (!data) {
        return [0, 0];
    }
    let fight = data.Fights[Recount.db.profile.CurDataSet];
    let score;
    if (fight && fight.Ovale && fight.OvaleMax) {
        score = fight.Ovale * 1000 / fight.OvaleMax;
    } else {
        score = 0;
    }
    if (num == 1) {
        return score;
    }
    return [score, undefined];
}
const TooltipFuncs = function(self: never, name: string) {
    GameTooltip.ClearLines();
    GameTooltip.AddLine(name);
}
export class OvaleRecountClass {
    constructor(private ovale: OvaleClass, private ovaleScore: OvaleScoreClass) {
        ovale.createModule("OvaleRecount", this.OnInitialize, this.OnDisable);
    }

    private OnInitialize = () => {
        if (Recount) {
            let aceLocale = AceLocale && AceLocale.GetLocale("Recount", true);
            if (!aceLocale) {
                aceLocale = setmetatable<LuaObj<string>>({}, {
                    __index: function (t, k) {
                        t[k] = k;
                        return k;
                    }
                });
            }
            Recount.AddModeTooltip(this.ovale.GetName(), DataModes, TooltipFuncs, undefined, undefined, undefined, undefined);

            this.ovaleScore.RegisterDamageMeter("OvaleRecount", this, this.ReceiveScore);
        }
    }
    private OnDisable = () => {
        this.ovaleScore.UnregisterDamageMeter("OvaleRecount");
    }
    private ReceiveScore = (name: string, guid: string, scored: number, scoreMax: number) => {
        if (Recount) {
            let source = Recount.db2.combatants[name];
            if (source) {
                Recount.AddAmount(source, this.ovale.GetName(), scored);
                Recount.AddAmount(source, `${this.ovale.GetName()}Max`, scoreMax);
            }
        }
    }
}
